machineconfiguration/v1alpha1/types_machineconfignode.go

```diff
@@ -107,6 +112,36 @@ type MachineConfigNodeStatus struct {
 	// This desired machine config has been compared to the current machine config and has been validated by the machine config operator as one that is valid and that exists.
 	// +kubebuilder:validation:Required
 	ConfigVersion MachineConfigNodeStatusMachineConfigVersion `json:"configVersion"`
+	// pinnedImageSetVersion describes the current and desired pinned image sets for this node.
+	// The current version is the generation of the pinned image set that has most recently been successfully pulled and pinned on this node.
+	// The desired version is the generation of the pinned image set that is targeted to be pulled and pinned on this node.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	PinnedImageSetVersion []MachineConfigNodeStatusPinnedImageSetVersion `json:"pinnedImageSetVersion,omitempty"`
```

Comment:
When you make use of this API, I suggest limiting updates by doing something like

on each node
1. read the set of pinnedimagesets that should be on the node
2. try to make actual state match desired state for up to day two minutes (API consumers will see the mismatch by looking at current state)
3. if actual state matches desired state, update once with conditions and state
4. if actual state does not match desired state, update once with conditions and state.

This gets the "normal" case down to a single update, keeps the normal case mismatch observable via the API, and keeps the latency of "why isn't working tolerable without having multiple updates for every node in the cluster at about the same time.

Reply:
great feedback :+1:

Reply:
> try to make actual state match desired state for up to day two minutes
just to be clear do work for at most 2 minutes then report status and retry if not complete.